guides/keras_nlp/neural_machine_translation.py

```diff
@@ -0,0 +1,509 @@
+
+eng_samples = [text_pair[0] for text_pair in train_pairs]
+eng_vocab = train_word_piece(eng_samples, ENG_VOCAB_SIZE, reserved_tokens)
+
+spa_samples = [text_pair[1] for text_pair in train_pairs]
+spa_vocab = train_word_piece(spa_samples, SPA_VOCAB_SIZE, reserved_tokens)
+
+"""
+Let's see some tokens!
+"""
+print("English Tokens: ", eng_vocab[100:110])
+print("Spanish Tokens: ", spa_vocab[100:110])
+
+"""
+Now, let's define the tokenizers. We will use the vocabularies obtained above as
+input to the tokenizers. We will define a maximum sequence length so that
+all sequences are padded to the same length, if the length of the sequence is
+less than the specified sequence length. Otherwise, the sequence is truncated.
+"""
+
+eng_tokenizer = keras_nlp.tokenizers.WordPieceTokenizer(
+    vocabulary=eng_vocab, lowercase=False, split_pattern=" ",
+)
+spa_tokenizer = keras_nlp.tokenizers.WordPieceTokenizer(
+    vocabulary=spa_vocab, lowercase=False, split_pattern=" ",
+)
+
+"""
+Let's try and tokenize a sample from our dataset! To verify whether the text has
+been tokenized correctly, we can also detokenize the list of tokens back to the
+original text.
+"""
+
+eng_input_ex = text_pairs[0][0]
+eng_tokens_ex = eng_tokenizer.tokenize(eng_input_ex)
+print("English sentence: ", eng_input_ex)
+print("Tokens: ", eng_tokens_ex)
+print("Recovered text after detokenizing: ", eng_tokenizer.detokenize(eng_tokens_ex))
+
+print()
+
+spa_input_ex = text_pairs[0][1]
+spa_tokens_ex = spa_tokenizer.tokenize(spa_input_ex)
+print("Spanish sentence: ", spa_input_ex)
+print("Tokens: ", spa_tokens_ex)
+print("Recovered text after detokenizing: ", spa_tokenizer.detokenize(spa_tokens_ex))
+
+"""
+## Format Datasets
+
+Next, we'll format our datasets.
+
+At each training step, the model will seek to predict target words N+1 (and beyond)
+using the source sentence and the target words 0 to N.
+
+As such, the training dataset will yield a tuple `(inputs, targets)`, where:
+
+- `inputs` is a dictionary with the keys `encoder_inputs` and `decoder_inputs`.
+`encoder_inputs` is the tokenized source sentence and `decoder_inputs` is the target sentence "so far",
+that is to say, the words 0 to N used to predict word N+1 (and beyond) in the target sentence.
+- `target` is the target sentence offset by one step:
+it provides the next words in the target sentence -- what the model will try to predict.
+
+We will add special tokens, `[START]` and `[END]`, to the input Spanish sentence
+after tokenizing the text.
+"""
+
+def normalize_length(tokens, max_sequence_length):
+    output_shape = tokens.shape.as_list()
+    output_shape[-1] = max_sequence_length
+    tokens = tokens.to_tensor(shape=output_shape)
+    return tokens
+
+def format_dataset(eng, spa):
+    batch_size = tf.shape(spa)[0]
+
+    eng = eng_tokenizer(eng)
+    spa = spa_tokenizer(spa)
+
+    # Add special tokens (`[START]` and `[END]`) to `spa`.
+    start_token_id_tensor = tf.expand_dims(
```

Comment:
Could you replace this line with `tf.fill((batch_size, 1), start_token)`, and ditto for end_token? Might be more readable